src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -88,10 +95,118 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
+                    flatten_logical_expressions(expression_in_disjunction, LogicalOperator::And);
+                for (const auto& expression_in_conjunction : inner_conjunction) {
+                  const auto evaluable_on_left_side =
+                      expression_evaluable_on_lqp(expression_in_conjunction, *join_node->left_input());
+                  const auto evaluable_on_right_side =
+                      expression_evaluable_on_lqp(expression_in_conjunction, *join_node->right_input());
+
+                  // We can only work with expressions that are specific to one side.
+                  if (evaluable_on_left_side && !evaluable_on_right_side && !aborted_left_side) {
+                    left_conjunction.emplace_back(expression_in_conjunction);
+                  }
+                  if (evaluable_on_right_side && !evaluable_on_left_side && !aborted_right_side) {
+                    right_conjunction.emplace_back(expression_in_conjunction);
+                  }
+                }
+
+                if (!left_conjunction.empty()) {
+                  // If we have found one or more predicates for the left side, connect them using AND and add them to
+                  // the disjunction that will be pushed to the left side.
+                  left_disjunction.emplace_back(inflate_logical_expressions(left_conjunction, LogicalOperator::And));
+                } else {
+                  // If, within the current expression_in_disjunction, we have not found a matching predicate for the
+                  // left side, all tuples for the left side qualify and it makes no sense to create a filter.
+                  aborted_left_side = true;
+                  left_disjunction.clear();
+                }
+                if (!right_conjunction.empty()) {
+                  right_disjunction.emplace_back(inflate_logical_expressions(right_conjunction, LogicalOperator::And));
+                } else {
+                  aborted_right_side = true;
+                  right_disjunction.clear();
+                }
+              }
+
+              const auto add_disjunction_if_beneficial =
+                  [&](const auto& disjunction, const auto& disjunction_input_node, auto& predicate_nodes) {
+                    if (disjunction.empty()) return;
+
+                    const auto expression = inflate_logical_expressions(disjunction, LogicalOperator::Or);
+                    const auto predicate_node = PredicateNode::make(expression, disjunction_input_node);
+
+                    // Determine the selectivity of the predicate if executed on disjunction_input_node
+                    const auto cardinality_in = _estimator->estimate_cardinality(disjunction_input_node);
+                    const auto cardinality_out = _estimator->estimate_cardinality(predicate_node);
+                    if (cardinality_out / cardinality_in > MAX_SELECTIVITY_FOR_PRE_JOIN_PREDICATE) return;
+                    predicate_node->set_left_input(nullptr);
```

Comment:
What do you do here?

Reply:
predicate_node was found to be beneficial. Add it to predicate_nodes so that _insert_nodes will
insert it as low as possible in the left/right input of the join. As predicate_nodes might have
more than one node, remove the input so that _insert_nodes can construct a proper LQP.

?

Reply:
Jep, that helps a lot.